### ISSUE TITLE
Collect usage statistics for various interfaces

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#2906]
I thought we had an issue already on that, but I cannot find it.  I would be nice if opt'ed-in users could submit anonymously information on what API functionality was used/failed/etc.  Might relate to #2866 (add reporting those statistics as a part of the issue/etc submission)

Existing libraries/helpers:
- https://github.com/popylar/popylar  track software use with Google Analytics
- https://pypi.org/project/usagestats/ might be the closest to the desired implementation

Worthwhile discussions
- [Django](https://github.com/django/deps/pull/31) - quick summary (if I got it right) of GDPR becoming a show stopper for opt-out

Other projects approaches
- AFNI the program checks for new releases, so could be used
- PsychoPy has "call home" ping url thingie with basic information (OS, version etc) so they collect stats

Other ideas on how we could "trigger" the submission
- as a part of the bug report/contact community (TODO: reference the issue) 
- part of `duecredit` collection
- if provides some additional service as 
  - report new version(s) available
  - list which other software might be of interest

Collections methodologies (should be configurable per project although @yarikoptic likes cached one big one instead of many small ones):
- immediate upon invocation
- cached collection which would then get submitted only after some time has passed since last submission or on cron, etc

What to collect
- UUID  (zenodo???)
- name of the package/software
  - possibly with indication that it was triggered by some other package on top (e.g. fmriprep calling out to nipype and then freesurfer etc)

**Comment by satra** [comment]
API on the server side:

- GET /version/{project}
- GET /ping
- POST /ping

**Comment by arokem** [comment]
Thanks for pinging me! Here is a summary of our experiences with popylar: 

As far as we can tell, the technical solution works well. The main challenges are social. Using Google Analytics as the data collection mechanism is probably a non-starter, because many users would consider the merge of their already collected data about browser/mail and so on, together with software usage patterns to be unacceptable. And I think that it wouldn't be a stretch. 

Users may have concerns even in the absence of the specific concerns related to GA. Opt-in seems to be the way to go, but likely to reduce your sample, and introduce biases. 

For Popylar in particular, I am wary of "heavy" solutions that require setting up persistent infrastructure (such as a database), because I think that this would raise a high barrier for use, but maybe you don't care about that for your use-case. I was considering trying Firebase as an intermediate solution, but haven't gotten around to it yet.

Given all these concerns, I haven't used Popylar for anything myself and it remains a work of conceptual art :-/

I do know that some other people have been considering using it. As I recall, @peircej showed this to someone who then started using it, but I am fuzzy on the details of that.

**Comment by peircej** [comment]
My method was to call a server with params (software version, platform, datetime) in the url, read them with PHP and store them with SQL. I used IP address to detect unique "users" (since GDPR I use a SHA of the IP). It's very simple to set up with a standard server setup, but it doesn't have the features of google analytics (e.g. geography mapping) or probably the resilience (we once lost several weeks of data when a database failed).

I like the concept of GA and Popylar for better graphing, richer analysis and being more robust but I found that popylar didn't work quite as easily as I wanted. Mostly I think this is just that the google API call is poorly documented and I couldn't work out what parameters were valid combinations. In particular if I want to store the software version and the operating system then I couldn't use multiple parameters in a single call. You would have to create an "event" called OS and another "event" called version etc. Not as easy as using GA for regular website visits. It all just felt a bit clunky. I tried to get some of the info uploaded using an agent info string but couldn't make GA use that either. 

As to concerns about privacy, note that GA provides the anonymous version and the non-anon version. If you turn off all the web marketing options then it is actually anonymous (can't be combined with the rest of google's data about the user). It's not anonymous when you add persistent tracking that (remarketing and advertising) and then you're legally bound by GDPR, assuming you have EU users. See https://support.google.com/analytics/answer/2700409
